Issue 121  install docs

kohr-h opened the pull request:
More extensive and updated installation docs, and a restructured "Getting started" section.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,397 @@
+The most popular ones are `Spyder`_ which works on all major platforms and can be installed through both ``conda`` and ``pip``, and `PyCharm`_ which can be integrated with any text editor of your choice, such as Emacs or Vim.
+
+
+Alternative 1: Installing a release using ``conda`` (recommended for users)
```
Have you tried this on a VM? Does it work on windows?

kohr-h replied:
I don't have a VM with a Windows more up to date than XP, so I was hoping that some Windows user could perhaps test it?^^

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,397 @@
+
+TL;DR
+-----
+- Download and install `Miniconda`_
```
Why not the full anaconda? If you need a guide like this you might as well get it.

kohr-h replied:
I think it's just waste of time and bandwidth. When installing ODL it will update the relevant packages anyway, so you can as well go with the small download.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,397 @@
+
+    $ conda install pytest
+
+
```
Lots of newlines

kohr-h replied:
Will be formatted away, but I'll delete one for you.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,397 @@
+Install a Python interpreter
+----------------------------
+Open a terminal and type ``python`` + Enter.
+If a Python prompt appears, you already have an interpreter installed and can skip this step.
```
Perhaps hint that they would exit, we dont want users to get stuck here :D

kohr-h replied:
Haha, I'll add a sentence.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,397 @@
+
+**On Windows:**
+
+Installers can be downloaded from `this link <https://www.python.org/downloads/windows/>`_.
```
We should be clear that anaconda is even more recommended on windows since installing numpy etc is hard as f.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,397 @@
+
+.. code-block:: bash
+
+   $ pip install --user odl
```
This will fail on windows with the python above (to my knowledge) quite sure the dependencies to build numpy are not available.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,397 @@
+
+Install ODL and its dependencies
+--------------------------------
+You need to `install pip`_ to be able to install ODL and its dependencies from the `Python Package Index`_ (PyPI).
```
http://stackoverflow.com/questions/4750806/how-do-i-install-pip-on-windows

> Good news! Python 3.4 (released March 2014) and Python 2.7.9 (released December 2014) ship with Pip. This is the best feature of any Python release.

So anyone installing python themselves should have pip, this is only a problem on old distros (like ubuntu 14). We should be clear in this, i.e. "You **may** need to"

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,397 @@
+
+**Extra dependencies**
+
+Additional dependencies can be specified in square brackets, separated by commas (no spaces!):
```
Be clear that these are not needed (see Frank trying to install all of them)

adler-j commented on doc/source/guide/numpy_guide.rst:
```diff
@@ -77,35 +77,35 @@ The fact that the ``x.ufunc.negative()`` interface is needed is a known issue wi
+   ...
+   ...     def _call(self, x):
+   ...         # The output of an Operator is automatically cast to an ODL vector
+   ...         return np.convolve(x, self.vector, mode='same')
```
It seems you have removed all my changes here. This should be fftconvolve

kohr-h replied:
This is only a small example, I'll keep `np.convolve`. Mostly because otherwise the results below are messed up, e.g.
```
File "/home/hkohr/git/odl/doc/source/guide/numpy_guide.rst", line 114, in numpy_guide.rst
Failed example:
    composed_op(x)
Expected:
    26.0
Got:
    26.000000000000004
```

adler-j replied:
ok!

kohr-h commented:
Okay, all comments fixed -- as far as I understood them. I also emphasized some things more by using `.. warning::` and `.. note::` directives, gives nice boxes in the final doc. I also tried to be more clear in places that were unclear before, and hopefully the structure is now easier to spot.

Merge?

kohr-h commented on doc/source/getting_started/about_odl.rst:
```diff
@@ -158,29 +138,24 @@ right-hand side is 0.14, so we choose 0.1.
-    0.240416305603
-    0.0991261822124
-    0.0408707719526
+    ...     print('{:.3}'.format(residual.norm()))
```
This was needed since Python 2 and 3 never print the same way automatically.

kohr-h commented on doc/source/getting_started/first_steps.rst:
```diff
@@ -206,23 +207,13 @@ Perhaps a bit better, but far from excellent.
-By identification, we see that the above problem can be written in this form if we let `f` be the indicator function on :math:`[0, 1]`, 
-:math:`g_1` be the squared l2 distance :math:`\| \cdot - g\|_2^2`, 
-:math:`g_2` be the norm :math:`\| \cdot \|_1`, 
-:math:`L_1` be the convolution operator and :math:`L_2` be the gradient operator.
```
Damnit, this got auto-merged in the wrong way. I'll fix it.

kohr-h replied:
> All of it is wrong, look further down.

What do you mean? Is d1c2a49 not getting it right? I don't see any duplicates.

adler-j commented on doc/source/getting_started/first_steps.rst:
```diff
@@ -232,27 +233,27 @@ e.g. `forward_backward_pd`, `chambolle_pock_solver`, etc in the ODL `examples/so
+
+   # Create proximals operators corresponding to the convex conjugate (cconj)
+   # of the l2 squared and l1 norms.
+   prox_cc_g = [odl.solvers.proximal_cconj_l2_squared(space, g=g),
```
This is where I meant that the merge failed, you fell back to using proximals

adler-j replied:
bumpty

adler-j commented on doc/source/getting_started/first_steps.rst:
```diff
@@ -232,27 +233,27 @@ e.g. `forward_backward_pd`, `chambolle_pock_solver`, etc in the ODL `examples/so
    # Solve using the Douglas-Rachford Primal-Dual method
    x = space.zero()
-   odl.solvers.douglas_rachford_pd(x, f, g_funcs, lin_ops,
+   odl.solvers.douglas_rachford_pd(x, prox_f, prox_cc_g, lin_ops,
```
Also here

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+TL;DR
+=====
+- Install `pip`_
+- Install `FFTW`_
```
Does this really need to be in the TLDR? I find it very un-important since it is only an optimization that rather advanced users could take advantage of, numpy has the same functionality.

kohr-h replied:
Not necessarily, but then they cannot use `pip install odl[all]`.

adler-j replied:
I'm frankly in favor of removing `pip install odl[all]`, people like Frank seem to think that they should do it while it is likely the last thing any normal user should do.

adler-j replied:
My argument is that users should not install pyfftw unless they:

* Know what they are doing
* Have already tried the numpy fft and find it to be slow

adler-j replied:
See https://github.com/odlgroup/odl/issues/665, I suggest we remove [all]

kohr-h replied:
Okay, that's the most robust solution.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+
+   $ pip install --user odl
+
+.. warning::
```
Perhaps a single warning (the one above) is enough.

kohr-h replied:
Yeah, it becomes a bit too much perhaps. Downgrade to note or remove?

adler-j replied:
Remove imo

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+
+
+------
+
```
.. alternative 3_

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+
+.. code-block:: bash
+
+   $ pip install --user --editable .
```
We changed from not having --user to having it, any reason?

adler-j replied:
Just wondering why we have different suggestions here

kohr-h replied:
Ah, but further up we're in a `conda` environment where it makes no sense to use `--user` since the default install location is user-writable.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+----------------
+.. code-block:: bash
+
+    $ conda install pytest pytest-cov pytest-pep8
```
Dislike this. Why not simply install the ODL dependencies? This way we can enforce dependencies etc. The way this is phrased atm would say "ok" if the user has pytest 2.x, but then the next step would fail since odl requires pytest>3.0.

kohr-h replied:
How would you get pytest 2.x with those instructions?

adler-j replied:
Well this line of code is independent of ODL, it doesnt mention it anywhere. Hence updating our dependencies would not make any difference here. Checking in code is indeed an option (that we should do), but it would only lead to users running this, getting an error and then having to run update anyway.

kohr-h replied:
Oh, but if it's in the dependencies then this step becomes obsolete altogether. The pytest stuff will be installed along with odl in the first place.

adler-j replied:
Valid point i guess. Then I'd also add matplotlib.

kohr-h replied:
Sure, maybe the `conda` package doesn't have to be that minimal.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+
+.. code-block:: bash
+
+    $ pip install --user pytest pytest-cov pytest-pep8
```
See above, we should do `pip install --user odl[testing]` and then perhaps add pytest-cov and pytest-pep8 to the testing requirements.

kohr-h replied:
Frankly we can skip the cov and pep8 stuff here. They're only really interesting to developers.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+
+   $ python -c "import odl; odl.test()"
+
+**anywhere but** in the top-level directory of an ODL clone.
```
This comment only applies for users that have installed from source, so perhaps move it to after the "if you have installed from source" thing?

kohr-h replied:
Not really, if you installed a release and run this in some (unrelated) clone, it will run the tests in the repo, not the installed ones.

adler-j replied:
Well then you have installed ODL from source *at some point*. I simply feel that this comment is not something people who use conda or pip need to read.

kohr-h replied:
Not even that. I suffices that you cloned the repo, it doesn't have to be installed.
Anyway, I'll make it a not and clarify that it only applies to source installations (move it).

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+Installation using ``conda``
+----------------------------
+.. note::
+    ``odlcuda`` is currently only available for Linux and Python 3.5.
```
only available in conda for

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+To calculate fast forward and backward projections for image reconstruction in X-ray tomography, install the `ASTRA tomography toolbox`_.
+ASTRA projectors are fully supported in ODL.
+
+You can try using the conda package, but we can give no guarantee that it works out of the box:
```
A link to their github for "for more help, please see..."

adler-j commented:
Requested some improvements for clarity, looks great overall but obviously needs to be tested against reality by new people.

kohr-h commented on doc/source/getting_started/first_steps.rst:
```diff
@@ -232,27 +233,27 @@ e.g. `forward_backward_pd`, `chambolle_pock_solver`, etc in the ODL `examples/so
                                    tau=tau, sigma=sigma, niter=100)
-   x.show('TV Douglas-Rachford', show=True)
-
+   x.show('TV Douglas-Rachford')
```
Damnit!

adler-j commented on doc/source/getting_started/about_odl.rst:
```diff
@@ -1,60 +1,53 @@
-`set`_ in that it can tell if an object belongs to it or not:
+A `Set` is the fundamental building block of ODL objects. It mirrors the mathematical concept of a `set`_ in that it can tell if an object belongs to it or not:
 
     >>> interv = odl.IntervalProd(0, 1)
```
Would be great with python typesetting of these, is that possible?

adler-j commented on doc/source/getting_started/about_odl.rst:
```diff
@@ -63,12 +56,9 @@ or :math:`\mathbb{C}`), and :math:`n` a positive integer. Example:
+The ODL implementation `FunctionSpace` covers only the cases :math:`\mathcal{Y} = \mathbb{R}` or :math:`\mathbb{C}` since the general case has large overlaps with `Operator`.
+Note that we do not make a distinction between different types of function spaces with respect to regularity, integrability etc. on an *abstract* level since there is no obvious way to check it.
 
 As linear spaces, function spaces support some interesting operations:
```
While reading the example i found this line:

>>> ratio_func = exp_plus_one / exp  # x -> exp(x) / (exp(x) + 1)

The comment has to be a misstake, right?

kohr-h replied:
Typo of course, good spot.

adler-j commented on doc/source/getting_started/about_odl.rst:
```diff
@@ -102,51 +89,44 @@ functions.
+    <class 'odl.discr.lp_discr.DiscreteLpElement'>
     >>> exp_discr.shape
     (5,)
 
```
Here printing its values would probably satisfy the readers curiosity :)

kohr-h replied:
Nothing simpler that that :-)

adler-j commented on doc/source/getting_started/about_odl.rst:
```diff
@@ -158,29 +138,24 @@ right-hand side is 0.14, so we choose 0.1.
+If we now exchange ``matrix_op`` and ``data`` with a tomographic projector and line integral data, not a single line of code in the reconstruction method changes since the operator interface is exactly the same.
 
 
 Further features
```
Here we should have doc linkage

adler-j commented on doc/source/getting_started/about_odl.rst:
```diff
@@ -158,29 +138,24 @@ right-hand side is 0.14, so we choose 0.1.
-  the Chambolle-Pock method (more to come...)
-* Standardized tests for the correctness of implementations of operators and spaces, e.g. does
-  the adjoint operator fulfill its defining relation?
+* Several solvers for variational inverse problems, ranging from simple gradient methods to the Chambolle-Pock method (more to come...)
```
Chambolle-pock isnt our most complicated. I'd say something about "nonsmooth convex solvers such as (list a few)" also remove "more to come".

adler-j commented on doc/source/getting_started/first_steps.rst:
```diff
@@ -1,8 +1,9 @@
-Welcome to the ODL getting started guide.
 This guide is intended to give you a simple introduction to ODL and how to work with it.
 If you need help with a specific function you should look at the `ODL API reference <https://odlgroup.github.io/odl/odl.html>`_.
 
```
A few lines down we write "whatver", spelling error.

adler-j replied:
Also found while reading this: We use conjugate_gradient_normal, but the operator is self adjoint and we could use conjugate_gradient. Do we want to fix this?

kohr-h replied:
Perhaps not. The normal equation variant is perhaps less common and appears more useful since it's more general. There's also the danger that users think `conjugate_gradient` makes no assumptions. I think this leads more into the right direction, even if it's not necessary.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+##############
+Installing ODL
+##############
+
```
A few words here or remove the "Introduction" title below. I find the style

Big Title
Small Title
text

To be quite hard to read and it feels like something was missed.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+1. Install a Python interpreter
+2. Install ODL and its dependencies
+3. (optional) Run the tests
+
```
Here we should be more clear that if the reader already has a python installation that they are happy with, they can skip the "install python" steps and go straight to the install ODL step

kohr-h replied:
I actually don't want to do that. Most Linux people have a Python interpreter, but I don't want them to miss the Anaconda recommendation. And they should know that they can skip things that they have installed themselves manually at some point.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+  .. code-block:: bash
+
+    $ conda create -c odlgroup -n odl-py35 python=3.5 odl odlcuda matplotlib pytest scikit-image spyder
+
```
Does the user need to activate the environment as well?

adler-j replied:
I would still add a line that implies it, even a fancy sounding:

> - Activate the conda environment and start working!

would be nice.

kohr-h replied:
Okay, convinced :-)

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+Installing ODL and its dependencies
+===================================
+You may need to `install pip`_ to be able to install ODL and its dependencies from the `Python Package Index`_ (PyPI).
+If running ``pip`` (alternatively: ``pip3``) an shows a help message, it is installed -- otherwise you need to install it first.
```
``pip2``/``pip3``

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,458 @@
+
+Possible choices:
+
+- ``all`` : Install with all extra dependencies.
```
I.m.o. get the #665 in here as well while at it.

kohr-h commented:
OK, I feel I have addressed all issues raised above. Maybe you want to check the conda env related thing @adler-j. Otherwise I'm quite happy with the overhaul, we should get this in asap.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,461 @@
+    to PATH in your /home/user/.bashrc ? [yes|no]
+    [no] >>> yes
+
+After opening a new terminal (for the changed ``PATH`` variable to take effect), you can run
```
I prefer `exec bash` to opening a new terminal

kohr-h replied:
Also, Ubuntu for example has its own shell "dash", so fixing bash would be hard-coding a non-obvious choice. And I don't know too much about shells on OSX.

adler-j replied:
I guess you are right. But perhaps go for "opening a new" to "restarting"? That kinda implies that `exec bash` is OK.

adler-j commented on doc/source/getting_started/installing.rst:
```diff
@@ -0,0 +1,461 @@
+
+    $ conda install pytest
+
+
```
Where did astra go here? They even have a conda package

kohr-h replied:
I'll add it with a link to the section all the way down.

adler-j commented on doc/source/guide/chambolle_pock_guide.rst:
```diff
@@ -0,0 +1,176 @@
+
+.. code-block:: python
+
+   >>> op_norm = 1.1 * odl.power_method_opnorm(L, maxiter=4, xstart=phantom)
```
Much better to use `xstart=data` here, both because it is more realistic and because it will give a better answer

kohr-h replied:
But that happens to work only in this case since the operator domain and range are the same. I find it misleading to use a range element as input to the norm estimator.
Or - if we want to do that - we could accept a range element and start with the adjiont evaluation. Would be a new issue, though.

adler-j commented:
Minor stuff, fix and merge.

kohr-h commented on doc/source/guide/chambolle_pock_guide.rst:
```diff
@@ -0,0 +1,176 @@
+
+Finally, we define the functional for the nonnegativity constraint, available as the functional `IndicatorNonnegativity`:
+
+    >>> g = odl.solvers.functional.IndicatorNonnegativity(space)
```
right

kohr-h commented on doc/source/guide/chambolle_pock_guide.rst:
```diff
@@ -0,0 +1,176 @@
+For the Chambolle-Pock method, there is one main rule that we can use:
+The product of the primal step :math:`\tau`, the dual step :math:`\sigma` and the squared operator norm :math:`\|L\|^2` has to be smaller than 1, :math:`\tau \sigma \|L\|^2 < 1`.
+Apart from this, there are no clear rules on how to select :math:`\tau` and :math:`\sigma` -- basically we're left with trial and error.
+We decide to pick them both equal to :math:`1 / \|L\|^2`.
```
oops
